### PR TITLE
Fix test-only goroutine leaks; add leak checker to end2end tests.

### PR DESCRIPTION
**Pull request by bradfitz**
Fixes grpc/grpc-go#528

### CODE REVIEW

**Comment by bradfitz** [comment]
@iamqizhao, this also has a few minor test/style cleanups and the --only_env flag for testing. Hope that's okay.

I'm off early for birthday activities now.

File: test/end2end_test.go [hunk 1]
```diff
@@ -317,19 +321,38 @@ func unixDialer(addr string, timeout time.Duration) (net.Conn, error) {
...
 	network  string // The type of network such as tcp, unix, etc.
 	dialer   func(addr string, timeout time.Duration) (net.Conn, error)
 	security string // The security protocol such as TLS, SSH, etc.
 }
 
+var (
+	tcpEnv        = env{name: "TCP", network: "tcp"}
+	tlsEnv        = env{name: "TLS", network: "tcp", security: "tls"}
+	unixSocketEnv = env{name: "Unix", network: "unix", dialer: unixDialer}
+	allEnv        = []env{tcpEnv, tlsEnv, unixSocketEnv}
```
Comment (iamqizhao) [inline test/end2end_test.go]: why throw away unix + tls?

Comment (iamqizhao) [inline test/end2end_test.go]: I think network and security are two orthogonal dimensions.

**Comment by iamqizhao** [comment]
@bradfitz 

Happy birthday and have fun! :)

File: test/end2end_test.go (hunk 1 above)
Comment (bradfitz) [inline test/end2end_test.go]: Oh, I missed that you even have "unix"+"tls" in the line noise that was line 329 before. :) I'll restore.

**Comment by bradfitz** [comment]
PTAL

File: test/end2end_test.go [hunk 2]
```diff
@@ -317,19 +321,39 @@ func unixDialer(addr string, timeout time.Duration) (net.Conn, error) {
...
 func listTestEnv() []env {
+	if *onlyEnv != "" {
+		for _, e := range allEnv {
+			if e.name == *onlyEnv {
+				return []env{e}
+			}
+		}
+		panic(fmt.Sprintf("invalid --only_env value %q", *onlyEnv))
+	}
 	if runtime.GOOS == "windows" {
```
Comment (iamqizhao) [inline test/end2end_test.go]: This should probably be placed before "if *only != "".

File: test/end2end_test.go [hunk 3]
```diff
@@ -317,19 +321,39 @@ func unixDialer(addr string, timeout time.Duration) (net.Conn, error) {
...
 
+var (
+	tcpClearEnv  = env{name: "tcp-clear", network: "tcp"}
+	tcpTLSEnv    = env{name: "tcp-tls", network: "tcp", security: "tls"}
+	unixClearEnv = env{name: "unix-clear", network: "unix", dialer: unixDialer}
+	unixTLSEnv   = env{name: "unix-tls", network: "unix", dialer: unixDialer, security: "tls"}
+	allEnv       = []env{tcpClearEnv, tcpTLSEnv, unixClearEnv, unixTLSEnv}
+)
+
+var onlyEnv = flag.String("only_env", "", "If non-empty, one of 'tcp-clear', 'tcp-tls', 'unix-clear', or 'unix-tls' to only run the tests for that environment. Empty means all.")
```
Comment (iamqizhao) [inline test/end2end_test.go]: in the description, mention this is only valid in non-windows platform?

File: test/end2end_test.go [hunk 4]
```diff
@@ -1253,3 +1312,64 @@ func testCompressOK(t *testing.T, e env) {
...
+// leakCheck snapshots the currently-running goroutines and returns a
+// function to be run at the end of tests to see whether any
+// goroutines leaked.
+func leakCheck(t testing.TB) func() {
+	orig := map[string]bool{}
+	for _, g := range interestingGoroutines() {
+		orig[g] = true
+	}
+	return func() {
+		t0 := time.Now()
```
Comment (iamqizhao) [inline test/end2end_test.go]: t0 -> t?

File: test/end2end_test.go [hunk 5]
```diff
@@ -1253,3 +1312,64 @@ func testCompressOK(t *testing.T, e env) {
...
+				if !orig[g] {
+					leaked = append(leaked, g)
+				}
+			}
+			if len(leaked) == 0 {
+				return
+			}
+			if time.Now().Before(t0.Add(5 * time.Second)) {
+				time.Sleep(50 * time.Millisecond)
+				continue
```
Comment (iamqizhao) [inline test/end2end_test.go]: What is the purpose we "continue" here? Can we just sleep 5 seconds and check once?

File: test/end2end_test.go (hunk 4 above)
Comment (bradfitz) [inline test/end2end_test.go]: t is used by the testing.TB already. t would shadow it. t0 is intentional and idiomatic: it's "time zero" (the start time).

File: test/end2end_test.go (hunk 2 above)
Comment (bradfitz) [inline test/end2end_test.go]: Windows users should still be able to filter. If they run a unix dialer environment and it fails, that's fine. Only developers will use this and they'll know.

File: test/end2end_test.go (hunk 3 above)
Comment (bradfitz) [inline test/end2end_test.go]: This should still work for Windows users. I think the name "unix-*" is clear enough that unix-* ones won't work on Windows.

File: test/end2end_test.go (hunk 4 above)
Comment (bradfitz) [inline test/end2end_test.go]: I renamed it to deadline and got rid of t0.

**Comment by bradfitz** [comment]
PTAL. Mirrored in Gerrit for better code review at: https://go-review.googlesource.com/#/c/19311/

**Pull request merged**